src/rust/engine/process_execution/src/remote.rs

```diff
@@ -1235,6 +1235,18 @@ mod tests {
       };
       let start_time = SystemTime::now();
       run_command_remote(mock_server.address(), execute_request).unwrap();
+//      mock_server.mock_responder.received_messages.lock().unwrap().iter().map(
+//        move |msg| match msg { (s,m,t) => println!("{:?}", t)}
+//      );
+//      let messages = mock_server.mock_responder.received_messages.lock().unwrap();
+//      assert!(messages.len() == 2);
+
+//      let (s, m, t) = messages.get(0).unwrap();
+      println!("{:?}",  mock_server.mock_responder.received_messages);
+//      println!("{:?}",  messages.get(1).unwrap().2.duration_since(messages.get(0).unwrap().2).unwrap());
```

Comment:
Currently
```
assert!(messages.len() == 2);
```
passes, but
```
println!("{:?}",  messages.get(1).unwrap().2.duration_since(messages.get(0).unwrap().2).unwrap());
```
prints
```
Duration { secs: 0, nanos: 1218000 }
```
which does not seem correct. we know line 1250 proves the backoff works, so not sure what I am missing here.

Reply:
Excitingly, this is pointing out a real bug, and I don't know why the test isn't currently failing.

The bug is that the logic currently reads:
```
If the response says the operation is not finished:
  Call GetOperation from the server
  Check whether we need to timeout or delay
  If we need to timeout, timeout
  If we need to delay, delay, then loop again
```

Where the logic should actually read:
```
If the response says the operation is not finished:
  Check whether we need to timeout or delay
  If we need to timeout, timeout
  If we need to delay, delay, then:
    Call GetOperation from the server, and loop
```

I'm not sure how I missed this in the original review!

If you move the code:
```
let grpc_result = map_grpc_result(operations_client.get().get_operation(&operation_request));
let operation = try_future!(grpc_result);
```

from where it currently is into the `and_then` block after the `Delay`, it will Just Work.

(We can then rewrite that block as:
```
future::done(map_grpc_result(operations_client.get().get_operation(&operation_request)))
                                .map(move |operation| future::Loop::Continue((operation, iter_num + 1)))
                                .to_boxed()
```
which is maybe a little cleaner?)

Reply:
Thanks for the pointer! Corrected